internal/impl/ollama/base_processor.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Licensed as a Redpanda Enterprise file under the Redpanda Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+// https://github.com/redpanda-data/connect/blob/main/licenses/rcl.md
+
+package ollama
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"io/fs"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path"
+	"time"
+
+	"github.com/dustin/go-humanize"
+	"github.com/ollama/ollama/api"
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+const (
+	bopFieldServerAddress = "server_address"
+	bopFieldModel         = "model"
+)
+
+func extractEmbeddedServer(fsys *service.FS, serverPath string) error {
+	_, err := fsys.Stat(serverPath)
+	if err == nil {
+		return nil
+	}
+	if errors.Is(err, fs.ErrNotExist) {
+		file, err := fsys.OpenFile(serverPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, os.ModePerm)
+		if err != nil {
+			return err
+		}
+		writer, isw := file.(io.Writer)
+		if !isw {
+			return errors.New("failed to write binary")
+		}
+		if _, err = writer.Write(embeddedServer); err != nil {
+			return err
+		}
+		return file.Close()
+	}
+	return err
+}
+
+type commandOutput struct {
+	logger *service.Logger
+	buffer []byte
+}
+
+func (co *commandOutput) Write(b []byte) (int, error) {
+	co.buffer = append(co.buffer, b...)
+	for {
+		idx := bytes.IndexByte(co.buffer, '\n')
+		if idx == -1 {
+			break
+		}
+		line := co.buffer[:idx]
+		if len(line) > 0 {
+			co.logger.Debugf("%s", line)
+		}
+		co.buffer = co.buffer[idx+1:]
+	}
+	return len(b), nil
+}
+
+func startServer(logger *service.Logger, fs *service.FS) (*exec.Cmd, error) {
+	serverPath, err := exec.LookPath("ollama")
+	if errors.Is(err, exec.ErrNotFound) {
+		if embeddedServer == nil {
+			return nil, errors.New("embedded ollama not supported for this platform - please specify a server_address")
+		}
+		serverPath = path.Join(os.TempDir(), "ollama")
```

Comment:
Small nit: I think we might want this path to be configurable in case people deal with readonly file systems.

Reply:
Good call - done!